File: treatment/src/main/java/com/hartwig/actin/treatment/ctc/CTCDatabaseReader.java
package com.hartwig.actin.treatment.ctc;

import java.io.IOException;
import java.util.List;
import java.util.Set;

import com.hartwig.actin.util.Paths;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

public final class CTCDatabaseReader {

    private static final Logger LOGGER = LogManager.getLogger(CTCDatabaseReader.class);

    private static final String CTC_DATABASE_TSV = "ctc_database.tsv";
    private static final String IGNORE_STUDIES_TSV = "ignore_studies.tsv";
    private static final String UNMAPPED_COHORTS_TSV = "unmapped_cohorts.tsv";

    @NotNull
    public static CTCDatabase read(@NotNull String ctcConfigDirectory) throws IOException {
        LOGGER.info("Reading CTC config from {}", ctcConfigDirectory);

        String basePath = Paths.forceTrailingFileSeparator(ctcConfigDirectory);

        return ImmutableCTCDatabase.builder()
                .entries(readCTCDatabaseEntries(basePath + CTC_DATABASE_TSV))
                .studyMETCsToIgnore(readIgnoreStudies(basePath + IGNORE_STUDIES_TSV))
                .unmappedCohortIds(readUnmappedCohorts(basePath + UNMAPPED_COHORTS_TSV))
                .build();
    }

    @NotNull
    private static List<CTCDatabaseEntry> readCTCDatabaseEntries(@NotNull String tsv) throws IOException {
        List<CTCDatabaseEntry> entries = CTCDatabaseEntryFile.read(tsv);
        LOGGER.info(" Read {} CTC database entries from {}", entries.size(), tsv);
        return entries;
    }

    @NotNull
    private static Set<String> readIgnoreStudies(@NotNull String tsv) throws IOException {
        Set<String> ignoreStudies = IgnoreStudiesFile.read(tsv);
        LOGGER.info(" Read {} study METCs to ignore from {}", ignoreStudies.size(), tsv);
        return ignoreStudies;
    }

    @NotNull
    private static Set<Integer> readUnmappedCohorts(@NotNull String tsv) throws IOException {
        Set<Integer> unmappedCohorts = UnmappedCohortFile.read(tsv);
        LOGGER.info(" Read {} unmapped cohorts from {}", unmappedCohorts.size(), tsv);
        return unmappedCohorts;
    }
}

File: treatment/src/main/kotlin/com/hartwig/actin/treatment/ctc/config/CTCDatabaseReader.kt
package com.hartwig.actin.treatment.ctc.config

import com.hartwig.actin.treatment.ctc.CTCDatabaseEvaluator
import com.hartwig.actin.util.Paths
import org.apache.logging.log4j.LogManager

object CTCDatabaseReader {

    private val LOGGER = LogManager.getLogger(CTCDatabaseReader::class.java)

    private const val CTC_DATABASE_TSV = "ctc_database.tsv"
    private const val IGNORE_STUDIES_TSV = "ignore_studies.tsv"
    private const val UNMAPPED_COHORTS_TSV = "unmapped_cohorts.tsv"

    fun read(ctcConfigDirectory: String): CTCDatabase {
        LOGGER.info("Reading CTC config from {}", ctcConfigDirectory)
        val basePath = Paths.forceTrailingFileSeparator(ctcConfigDirectory)
        val ctcDatabase = CTCDatabase(
            entries = readCTCDatabaseEntries(basePath + CTC_DATABASE_TSV),
            studyMETCsToIgnore = readIgnoreStudies(basePath + IGNORE_STUDIES_TSV),
            unmappedCohortIds = readUnmappedCohorts(basePath + UNMAPPED_COHORTS_TSV)
        )

        LOGGER.info("Evaluating usage of CTC database configuration")
        val ctcDatabaseEvaluator = CTCDatabaseEvaluator(ctcDatabase)
        ctcDatabaseEvaluator.evaluateDatabaseConfiguration()

        return ctcDatabase
    }

    private fun readCTCDatabaseEntries(tsv: String): List<CTCDatabaseEntry> {
        val entries = CTCDatabaseEntryFile.read(tsv)
        LOGGER.info(" Read {} CTC database entries from {}", entries.size, tsv)
        return entries
    }

    private fun readIgnoreStudies(tsv: String): Set<String> {
        val ignoreStudies = IgnoreStudiesFile.read(tsv)
        LOGGER.info(" Read {} study METCs to ignore from {}", ignoreStudies.size, tsv)
        return ignoreStudies
    }

    private fun readUnmappedCohorts(tsv: String): Set<Int> {
        val unmappedCohorts = UnmappedCohortFile.read(tsv)
        LOGGER.info(" Read {} unmapped cohorts from {}", unmappedCohorts.size, tsv)
        return unmappedCohorts
    }
}